### DISCUSSION TITLE
ConcurrentT Waiting

### DISCUSSION THREAD

**Discussion by ggwpez** [hyperledger-labs/go-perun#177] [Ideas] [answered]
We currently use strings for stages which is error prone, even when there is only one stage.  
The simplest example looks like this:
```go
ct := pkgtest.NewConcurrent(t)
ct.Stage("fund", func(t pkgtest.ConcT) {
	…
})
…
ct.Wait("fund")
```
I propose it to enable the following usage:  
```go
stage := pkgtest.NewConcurrent(t).Stage("fund", func(t pkgtest.ConcT) {
	…
})
…
stage.Wait()
```
Anther possibility would be to add a `WaitAll()`.   
I dont have enough knowledge about the code, so i created a discussion instead of an issue.  
What do you think @RmbRT ?

**Comment by RmbRT** [comment, marked as answer]
Sometimes, we need to wait for stages that were started somewhere else. For that, we'd need to keep the `Wait("name")` method. Since `Stage` is already a blocking function that implicitly waits for the stage, the above example is technically incorrect. However, if the stage is run inside a goroutine, we need to manually await it when required. For that case, it would be helpful to create a `GoStage` series of functions that returns an awaitable reference to the stage, but runs the stage itself in a goroutine.
I'm not sure about `WaitAll`, it's kind of risky because it doesn't wait for stages that have not yet been initiated. It's still safest to use `ct.Wait("stage1", "stage2", ...)` regarding races. An optional whitelist of stage names could also be configured upon creation of the `ConcurrentT`, but that's not simple to maintain if you have a dynamic amount of stages (e.g. stages in loops).

**Reply by ggwpez** [reply]
Ok thanks for the answer, i just saw that the example does not start the stage in a `go` :facepalm:  
We can then open an issue for your `GoStage` idea.
